### ISSUE TITLE
DataLoader causes scalar indexing on GPU in Flux v0.13.0 (regression)

### ISSUE THREAD

**Issue by oschulz** [FluxML/Flux.jl#1935]
With CUDA v3.8.5 and Flux v0.12.10 this runs fine

```julia
using Flux
using CUDA

x = rand(10, 1000) |> gpu;
y = rand(1000) |> gpu;

dataloader = Flux.Data.DataLoader((x, y), batchsize=512, shuffle=true);

iterate(dataloader)
```

But with CUDA v3.8.5 and Flux v0.13.0 I reproduceably get

```
ERROR: LoadError: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
...
```

The output of `iterate(dataloader)` seems correct, though.

Initially discovered by @lmh91

**Comment by darsnack** [comment]
The issue occurs when `shuffle = true` specifically. Normally, the batch iterator would call `getobs(data, m:n)` where the range `m:n` is contiguous, so it doesn't trigger scalar indexing. With shuffling turned on, we call `getobs(data, shuffled_idx[m:n])` which triggers the error.

It's still unclear to me why this wasn't an issue in Flux@0.12.10, since the sequence of operations appears to be the same.

**Comment by darsnack** [comment]
Ah, it's cause the shuffling operation, `shuffleobs`, has a "fast path" which returns a view instead of keeping the original data + shuffled indices. Indexing a view of a `CuArray` with a vector of indices triggers scalar indexing. This is a known issue on the CUDA side (I can't seem to find the issue link), but I don't think there is any intention of fixing it soon.

Probably we need to stop `CuArray`s from taking the "fast path."

**Comment by CarloLucibello** [comment]
Was just commenting something similar to what @darsnack said
```julia
julia> using CUDA,Flux,MLUtils

julia> y = rand(5) |> gpu
5-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 0.15889172
 0.5090761
 0.45227328
 0.86806625
 0.11143296

julia> v = view(y,[5,4,3,2,1])
5-element view(::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, [5, 4, 3, 2, 1]) with eltype Float32:
 0.11143296
 0.86806625
 0.45227328
 0.5090761
 0.15889172

julia> y[1:2]
2-element CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}:
 0.15889172
 0.5090761

julia> v[1:2]
ERROR: Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore are only permitted from the REPL for prototyping purposes.
If you did intend to index this array, annotate the caller with @allowscalar.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] assertscalar(op::String)
    @ GPUArrays ~/.julia/packages/GPUArrays/Zecv7/src/host/indexing.jl:53
  [3] getindex
    @ ~/.julia/packages/GPUArrays/Zecv7/src/host/indexing.jl:86 [inlined]
  [4] reindex
    @ ./subarray.jl:254 [inlined]
  [5] getindex
    @ ./subarray.jl:276 [inlined]
  [6] macro expansion
    @ ./multidimensional.jl:867 [inlined]
  [7] macro expansion
    @ ./cartesian.jl:64 [inlined]
  [8] _unsafe_getindex!
    @ ./multidimensional.jl:862 [inlined]
  [9] _unsafe_getindex(#unused#::IndexCartesian, A::SubArray{Float32, 1, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}, false}, I::UnitRange{Int64})
    @ Base ./multidimensional.jl:853
 [10] _getindex
    @ ./multidimensional.jl:839 [inlined]
 [11] getindex(A::SubArray{Float32, 1, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}, false}, I::UnitRange{Int64})
    @ Base ./abstractarray.jl:1218
 [12] top-level scope
    @ REPL[22]:1
 [13] top-level scope
    @ ~/.julia/packages/CUDA/5jdFl/src/initialization.jl:52
```

**Comment by oschulz** [comment]
Specializing obsview for GPU arrays seems to fix it. Would this be acceptable?

```julia
function Flux.obsview(A::AbstractGPUArray{T,N}, idx) where {T,N}
    I = ntuple(_ -> :, N-1)
    return getindex(A, I..., idx)
end
```

**Comment by oschulz** [comment]
> Specializing obsview for GPU arrays seems to fix it. Would this be acceptable?

Ah, sorry, saw JuliaML/MLUtils.jl#73 too late :-)

**Comment by CarloLucibello** [comment]
> Specializing obsview for GPU arrays seems to fix it. Would this be acceptable?

Specialization on AbstractGPUArrays doesn't capture wrapped cuarrays
```julia
julia> x = CUDA.zeros(5)'
1×5 adjoint(::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}) with eltype Float32:
 0.0  0.0  0.0  0.0  0.0

julia> x isa CUDA.AbstractGPUArray
false
```

**Comment by CarloLucibello** [comment]
This is fixed with the just-released MLUtils v0.2.3
